File: apps/ios/Sources/Core/Identity/IdentityAttempt.swift
import Foundation

enum IdentityIntent: Sendable { case login, link }
struct IdentityScope: Equatable, Sendable {
    let environment: String
    let sessionEpoch: UUID
    let accountID: String?
}
struct IdentityTicket: Equatable, Sendable {
    fileprivate let operation: UUID
    let scope: IdentityScope
    let intent: IdentityIntent
}
// OS writer binds protected persistent epoch before presenting Apple/browser UI.
// Server /me determines SOOP readiness; provider credential alone never does.
struct IdentityAttempt: Sendable {
    private var scope: IdentityScope?
    private var pending: IdentityTicket?
    mutating func bind(_ value: IdentityScope) {
        if scope != value { pending = nil; scope = value }
    }
    mutating func begin(_ intent: IdentityIntent) -> IdentityTicket? {
        guard let scope, intent == .login ? scope.accountID == nil : scope.accountID != nil else { return nil }
        let ticket = IdentityTicket(operation: UUID(), scope: scope, intent: intent)
        pending = ticket
        return ticket
    }
    func current(_ ticket: IdentityTicket) -> Bool { pending == ticket && scope == ticket.scope }
    mutating func consume(_ ticket: IdentityTicket) -> Bool {
        guard current(ticket) else { return false }
        pending = nil
        return true
    }
    mutating func cancel() { pending = nil }
}

File: apps/android/app/src/main/java/chat/rogi/rogichat/core/identity/IdentityAttempt.kt
package chat.rogi.rogichat.core.identity

import java.util.UUID

enum class IdentityIntent { LOGIN, LINK }
data class IdentityScope(val environment: String, val sessionEpoch: UUID, val accountId: String?)
class IdentityTicket internal constructor(val scope: IdentityScope, val intent: IdentityIntent)

/** Scope comes from protected persistent epoch; provider identity never grants SOOP chat access. */
class IdentityAttempt {
    private var scope: IdentityScope? = null
    private var pending: IdentityTicket? = null
    fun bind(value: IdentityScope) { if (scope != value) { pending = null; scope = value } }
    fun begin(intent: IdentityIntent): IdentityTicket? {
        val scope = scope ?: return null
        if (if (intent == IdentityIntent.LOGIN) scope.accountId != null else scope.accountId == null) return null
        return IdentityTicket(scope, intent).also { pending = it }
    }
    fun current(ticket: IdentityTicket) = pending === ticket && scope == ticket.scope
    fun consume(ticket: IdentityTicket): Boolean {
        if (!current(ticket)) return false
        pending = null
        return true
    }
    fun cancel() { pending = null }
}
